bin/sdmerge

```diff
@@ -1,148 +1,228 @@
 <REPO-NAME>   the repo name, for example: quasar-analytics/quasar
 <PR-NUMBER>   the pull request number: 2030
 
-Example: sdmerge quasar-analytics/quasar 2030
+Example: sdmerge quasar-analytics/quasar 2030""".format(sys.argv[0])
+    exit(1)
+
 
-EOF
-}
+def infer_repo_slug():
+    out = subprocess.check_output(["git", "remote", "-v"])
+    upstream_fetch_regex = "^" + re.escape(UPSTREAM) + r"\t(?:git@github\.com:|https://github\.com/)(.+?)(?:.git)? \(fetch\)$"
+    match = re.compile(upstream_fetch_regex, re.MULTILINE).search(out)
+    if match is None:
+        print "error: unable to determine current working repository (do you have a remote named '%s'?)" % UPSTREAM
+        exit(1)
+    else:
+        return match.group(1)
 
-# if no args are passed in print usage
-[ $# -eq 0 ] && usage
 
-# command line parsing logic
-while getopts ":h" opt; do
-  case $opt in
-    h | *)
-      usage
-      exit 1
-      ;;
-  esac
-done
+def call_github_api(url, token):
+    conn = httplib.HTTPSConnection("api.github.com")
 
-mainwork $*
+    headers = {"User-Agent": "sdmerge"}  # user agent is required by github api
+    if token is not None:
+        headers.update({"Authorization": "token " + token})
+
+    conn.request("GET", url, headers=headers)
+    response = conn.getresponse()
+    if response.status != 200:
+        if response.status == 403 and token is None:
+            print "error: set GITHUB_TOKEN env variable to access private repositories"
+        elif response.status == 404:
+            print "error: pull request with this number does not exist"
+            if token is None:
+                print "error: if you are trying to access private repository make sure that GITHUB_TOKEN is set"
+        else:
+            print "error: invalid response: ", response.status, response.reason
+            print response.read()
+        conn.close()
+        exit(1)
+    else:
+        raw_data = response.read()
+        conn.close()
+        return json.loads(raw_data)
+
+
+def get_pr_metadata():
+    github_token = os.environ.get("GITHUB_TOKEN")
+    endpoint = "/repos/%s/pulls/%s" % (repo_slug, pull_request_number)
+    return call_github_api(endpoint, github_token)
+
+
+def read_sbt_version():
+    with open("version.sbt") as file:
+        return re.match(r"version in ThisBuild := \"(.+)\"", file.read()).group(1)
+
+
+def read_npm_version():
+    with open("package.json") as file:
+        return json.load(file)["version"].replace("v", "")
+
+
+def save_sbt_version(version):
+    with open("version.sbt", "w") as file:
+        file.write("version in ThisBuild := \"%s\"" % version)
+
+
+def save_npm_version(version):
+    with open("package.json") as file:
+        content = file.read()
+
+    # sadly parsing file as json, replacing the value and saving it back reorders everything, doing a big mess
+    new_content = re.sub("(?<=\"version\": \").+?(?=\")", "v" + version, content)
```

Comment:
I was wondering how you were going to handle this issue. 😊 This seems as reasonable a solution as any.

Reply:
Good thing is that this is not the case with jq, so at least dependency bumping script works as it should. But with std python probably there is no much better way, except for implementing a parser.

Reply:
`json.loads(content, object_pairs_hook=OrderedDict)` preserves the field ordering, did you try that?

Reply:
Nope. Looks pretty obvious, now that I seen that, not sure why I didn't came across that when looking for a solution. I made a PR with this improvement :smile:  https://github.com/slamdata/devtools/pull/32